src/envoy/http/service_control/handler_impl.cc

```diff
@@ -143,11 +145,20 @@ void ServiceControlHandlerImpl::prepareReportRequest(
     ::espv2::api_proxy::service_control::ReportRequestInfo& info) {
   fillOperationInfo(info);
 
-  // Report: not to send api-key if invalid or service is not enabled.
-  if (check_response_info_.error_type == ScResponseErrorType::API_KEY_INVALID ||
-      check_response_info_.error_type ==
-          ScResponseErrorType::SERVICE_NOT_ACTIVATED) {
-    info.api_key.clear();
+  // This logic determines if consumer metric is sent in report based on the
+  // validity of the API Key.
+  if (on_check_done_called_) {
+    if (check_response_info_.error_type ==
+        ScResponseErrorType::API_KEY_INVALID) {
+      info.api_key_state = ApiKeyState::INVALID;
+    } else if (check_response_info_.error_type ==
+               ScResponseErrorType::SERVICE_NOT_ACTIVATED) {
+      info.api_key_state = ApiKeyState::NOT_ENABLED;
+    } else {
+      info.api_key_state = ApiKeyState::VERIFIED;
```

Comment:
How about Check returned with no-ok status?  network failure status?  I think the default check_respond_info_check_type is NOT_SPECIFIED.   We may have to handle that.

Reply:
Good catch. Ok, now the logic is getting too complicated for `HandlerImpl`. 

I agree with your point that we should do the conversion in `ClientCache`, I changed the code.